[2d] Ensure gradient clear out pending AsyncCollectiveTensor in FSDP Extension

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116573
* #116559
* #116426
* #116244
* __->__ #116122

As titled, this PR adds gradient hook to the FSDP DTensor extension, to check if there's gradients that are AsyncCollectiveTensors, if there're some, we eagerly wait there.

This is needed because sometimes the parameter's gradient might still pending with AsyncCollectiveTensor, if we directly feed them to FSDP then FSDP would use the ACT's storage to do reduce_scatter, which is wrong.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[2d] add gradient hook to FSDP extension

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] Ensure param/grad sharding layout consistency in FSDP extension"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] Ensure param/grad sharding layout consistency in FSDP extension"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] Ensure param/grad sharding layout consistency in FSDP extension"


As titled, this PR adds gradient hook to the FSDP DTensor extension,
given the fact that FlatParam FSDP assumes same sharding layout after
backward pass, here we also ensure the DTensor gradients follow the same
sharding layout

[ghstack-poisoned]

wanchaol commented on test/distributed/fsdp/test_fsdp_tp_integration.py:
```diff
@@ -314,6 +351,78 @@ def test_fsdp_tp_integration(self, tensor_parallel_size, cpu_offload):
         self.assertEqual(fsdp_out, tp_fsdp_out)
 
+    @skip_if_lt_x_gpu(4)
+    def test_fsdp_tp_gradient_layout(self):
```
@awgu I added some tests on this change so this PR should be ready for review too :)

awgu replied:
Thanks! I will review soon.

awgu commented on test/distributed/fsdp/test_fsdp_tp_integration.py:
```diff
@@ -314,6 +351,78 @@ def test_fsdp_tp_integration(self, tensor_parallel_size, cpu_offload):
+                "mlp.net2": RowwiseParallel(output_layouts=Shard(0)),
+            },
+        )
+        distribute_rmsnorm(tp_model.mlp_norm, mesh_1d)
```
IIUC, this means the RMSNorm runs as "sequence parallel"?

wanchaol replied:
Yep, this is sequence parallel layernorm/rmsnorm, the backward would produce partial dtensor, and the parameter would be replicated dtensor

awgu reviewed (approved):
LGTM!

fduwjj reviewed (approved):
LGTM

wanchaol committed:
Update on "[2d] Ensure param/grad sharding layout consistency in FSDP extension"


As titled, this PR adds gradient hook to the FSDP DTensor extension,
given the fact that FlatParam FSDP assumes same sharding layout after
backward pass, here we also ensure the DTensor gradients follow the same
sharding layout

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] Ensure param/grad sharding layout consistency in FSDP extension"


As titled, this PR adds gradient hook to the FSDP DTensor extension,
given the fact that FlatParam FSDP assumes same sharding layout after
backward pass, here we also ensure the DTensor gradients follow the same
sharding layout

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] Ensure param/grad sharding layout consistency in FSDP extension"


As titled, this PR adds gradient hook to the FSDP DTensor extension,
given the fact that FlatParam FSDP assumes same sharding layout after
backward pass, here we also ensure the DTensor gradients follow the same
sharding layout

[ghstack-poisoned]

wanchaol commented:
changed the purpose of this PR a bit to make the gradient hook to eagerly wait for the ACT.

awgu commented on torch/distributed/tensor/parallel/_data_parallel_utils.py:
```diff
@@ -1,24 +1,34 @@
 
+def sync_grad_hook(grad):
+    if isinstance(grad, AsyncCollectiveTensor):
+        grad.wait()
```
@wanchaol Why do we discard the return value of `grad.wait()`, which unwraps the ACT? I.e., why not `grad = grad.wait()`?

wanchaol replied:
Good catch! I think we should unwrap yes, it's a bug. Could u help submit a fix on this?

awgu replied:
Oh wait, this looks like it was fixed on main already.
